distribution/lib/Standard/Base/0.0.0-dev/src/Data/Json/Extensions.enso

```diff
@@ -1,136 +1,137 @@
-
-   > Example
-     Converting a dataflow error to JSON.
-
-         import Standard.Examples
-
-         example_to_json = Examples.throw_error.to_json
-Error.to_json : Json
-Error.to_json self =
-    error_type = ["type", "Error"]
+            type_name = Meta.get_qualified_type_name self . split '.' . at -2
+            JS_Object.from_pairs [["type", type_name], ["constructor", m.name]]
+        _ -> Error.throw ("Cannot convert " + self.to_text + " to JSON")
+
+## Converts the given value to a JSON serializable object.
+   For Nothing, booleans, numbers and strings, this is the value itself.
+   For arrays or vectors, the elements are converted recursively.
+   For atoms, the object is converted to a JS_Object.
+Error.to_js_object : JS_Object
+Error.to_js_object self =
     caught = self.catch
-    error_content = ["content", caught.to_json]
+    error_type = ["type", "Error"]
+    error_content = ["content", caught]
     error_message = ["message", caught.to_display_text]
-    Json.from_pairs [error_type, error_content, error_message]
-
-## Text to JSON conversion.
-
-   > Example
-     Convert the text "Hello World!" to JSON.
-
-         "Hello World!".to_json
-   > Example
-     Convert the text "cześć" to JSON.
-
-         "cześć".to_json
-Text.to_json : Json
-Text.to_json self = Json.String self
-
-## Method used by object builders to convert a value into a valid JSON key.
-
-   > Example
-     Ensure that the text "foo" is a JSON key.
-         "foo".to_json_key
-Text.to_json_key : Text
-Text.to_json_key self = self
-
-## Convert a boolean to JSON.
-
-   > Example
-     Convert `True` to JSON.
-         True.to_json
-Boolean.to_json : Json
-Boolean.to_json self = Json.Boolean self
-
-## Convert `Nothing` to JSON.
-
-   > Example
-     Convert `Nothing` to JSON.
-         Nothing.to_json
-Nothing.to_json : Json
-Nothing.to_json self = Json.Null
-
-
-## Number to JSON conversion.
-
-   > Example
-     Convert the number 8 to JSON.
-
-         8.to_json
-Number.to_json : Json
-Number.to_json self = Json.Number self
-
-## Vector to JSON conversion.
-
-   > Example
-     Convert a vector of numbers to JSON.
-
-         [1, 2, 3].to_json
-Vector.to_json : Json
-Vector.to_json self = Json.Array (self.map .to_json)
-
-## A Locale to Json conversion
-
-   > Example
-     Convert the default locale to JSON.
-
-         import Standard.Base.Data.Locale.Locale
+    JS_Object.from_pairs [error_type, error_content, error_message]
 
-         example_to_json = Locale.default.to_json
-Locale.to_json : Json
-Locale.to_json self =
-    b = Vector.new_builder
+## Converts the given value to a JSON serializable object.
+   Custom serialization for Locale, serializes the language, country and variant.
+Locale.to_js_object : JS_Object
+Locale.to_js_object self =
+    b = Vector.new_builder 5
     b.append ["type", "Locale"]
-    if self.language.is_nothing.not then b.append ["language", self.language]
+    b.append ["constructor", "new"]
```

Comment:
Why is the `constructor` a method name? Seems surprising.
Do we even need the `constructor` here? I guess it makes sense to be consistent with Atom serialization, but we put a method name instead of true constructor so we are not truly consistent anyway.

Reply:
The idea is to make it, so we had the information to deserialise.
Originally had fully qualified type names but decided against it for this version.